utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -60,31 +56,93 @@ pub struct TryRuntimeCmd {
 	)]
 	pub wasm_method: WasmExecutionMethod,
+
+	/// The state to use to run the migration.
+	#[structopt(subcommand)]
+	pub state: State,
 }
 
 /// The state to use for a migration dry-run.
-#[derive(Debug)]
+#[derive(Debug, structopt::StructOpt)]
 pub enum State {
-	/// A snapshot. Inner value is a file path.
-	Snap(String),
+	/// Use a snapshot as state to run the migration.
+	Snap {
+		#[structopt(flatten)]
+		cache_params: CacheParams,
+	},
+
+	/// Use a live chain to run the migration.
+	Live {
+		/// An optional cache file to WRITE to. Not cached if set to `None`.
+		#[structopt(short, long)]
+		cache_file: Option<CacheParams>,
+
+		/// The block hash at which to connect.
+		/// Will be latest finalized head if not provided.
+		#[structopt(short, long, multiple = false, parse(try_from_str = parse_hash))]
+		block_at: Option<String>,
+
+		/// The modules to scrape. If empty, entire chain state will be scraped.
+		#[structopt(short, long, require_delimiter = true)]
+		modules: Option<Vec<String>>,
+
+		/// The url to connect to.
+		#[structopt(default_value = "http://localhost:9933", parse(try_from_str = parse_url))]
+		url: String,
+	},
+}
 
-	/// A live chain. Inner value is the HTTP uri.
-	Live(String),
+fn parse_hash(block_number: &str) -> Result<String, String> {
+	let block_number = if block_number.starts_with("0x") {
+		&block_number[2..]
+	} else {
+		block_number
+	};
+
+	if let Some(pos) = block_number.chars().position(|c| !c.is_ascii_hexdigit()) {
+		Err(format!(
+			"Expected block hash, found illegal hex character at position: {}",
+			2 + pos,
+		))
+	} else {
+		Ok(block_number.into())
+	}
+}
+
+fn parse_url(s: &str) -> Result<String, &'static str> {
+	if s.starts_with("http://") {
+		// could use Url crate as well, but lets keep it simple for now.
+		Ok(s.to_string())
+	} else {
+		Err("not a valid HTTP url: must start with 'http://'")
+	}
 }
 
-impl FromStr for State {
+#[derive(Debug, structopt::StructOpt)]
+pub struct CacheParams {
+	/// The directory of the snapshot.
+	#[structopt(short, long, default_value = ".")]
+	directory: String,
+
+	/// The file name of the snapshot.
+	#[structopt(default_value = "CACHE")]
+	file_name: String,
+}
+
+impl FromStr for CacheParams {
 	type Err = &'static str;
 	fn from_str(s: &str) -> Result<Self, Self::Err> {
-		match s.get(..7) {
-			// could use Url crate as well, but lets keep it simple for now.
-			Some("http://") => Ok(State::Live(s.to_string())),
-			Some("file://") => s
-				.split("//")
-				.collect::<Vec<_>>()
-				.get(1)
-				.map(|s| State::Snap(s.to_string()))
-				.ok_or("invalid file URI"),
-			_ => Err("invalid format. Must be a valid HTTP or File URI"),
+		let p: PathBuf = s.parse().map_err(|_| "invalid path")?;
+		let parent = p.parent();
+		let file_name = p.file_name();
+
+		match file_name {
+			Some(file_name) => Ok(Self {
+				// TODO: maybe don't use to_string_lossy here
+				directory: parent.map(|p| p.to_string_lossy().into()).unwrap_or(".".to_string()),
```

Comment:
I think it's a bit dangerous to let an invalid path through here. A bit contrived perhaps, but if `file_name` is present both in `.` and whatever path I thought I gave then something bad could happen?

Reply:
actually there's a TODO in `CacheConfig` as 

```
	// TODO: I could mix these two into one filed, but I think separate is better bc one can be
	// configurable while one not.
```

I also thought about the dangers. You are potentially giving write primitive to an external cli input. Technically, a malicious input can overwrite some critical file for example. But the thing is, no one is going to expose this functionality in the wild or anything. It will all be within the trust boundary, so I think it is fine.

Reply:
That being said, sanitizing the path is a good idea anyway.

Reply:
I think it might be a good idea to use `PathBuf` internally. This way we can at least be sure that no chars will be dropped during the string conversion. And we can expose the `PathBuf` directly as a structopt parameter, which would remove the need for manual conversions at all.

I also think it might would be better to combine directory and filename into one field. Internally it is combined to a single `PathBuf` anyway. If we still want the directory and path to be set separately, we can still do this on the cli level, but then it also needs to be validated that the file_name is always just a file and doesn't contain directory parts. Otherwise the directory parameter would be redundant.

Reply:
@kianenigma Do you want me to combine the paths into a single `PathBuf` for this PR or should that be handled separately?

Reply:
I prefer not having major changes anymore once the PR is approves, but given the simplicity of this, whichever your find better, both fine by me.

Reply:
Okay, I'm gonna prepare a separate PR then.